### PR TITLE
Gracefully flush chunks when ingestor exits

### PR DESCRIPTION
**Pull request by tomwilkie**

### CODE REVIEW

File: frankenstein/cmd/frankenstein/main.go [hunk 1]
```diff
@@ -86,7 +86,9 @@ func main() {
 	case distributor:
 		setupDistributor(consul, consulPrefix, chunkStore, remoteTimeout)
 	case ingestor:
-		setupIngestor(consul, consulPrefix, chunkStore, listenPort)
+		ingestor := setupIngestor(consul, consulPrefix, chunkStore, listenPort)
+		defer deleteIngestorConfigFromConsul(consul, consulPrefix)
+		defer ingestor.Stop()
```
Comment (juliusv) [inline frankenstein/cmd/frankenstein/main.go:92]: Doesn't this need a signal handler? Like in https://github.com/prometheus/prometheus/blob/master/cmd/prometheus/main.go#L193-L202

When stopping a Go program with SIGTERM (not sure what k8s sends) without a signal handler, the deferred functions are not executed.

Comment (tomwilkie) [inline frankenstein/cmd/frankenstein/main.go:92]: Yes it will!

Comment (juliusv) [inline frankenstein/cmd/frankenstein/main.go:92]: You mean they'll be run? Odd cause when I try it locally with this and send SIGTERM, it doesn't print "EXITING":

``` go
package main

import "fmt"

func foo() {
    fmt.Println("EXITING")
}

func main() {
    defer foo()
    for {
    }
}
```

What's different here?

Comment (tomwilkie) [inline frankenstein/cmd/frankenstein/main.go:92]: Sorry wasn't clear - I mean "yes it will" need a signal handler.

Comment (juliusv) [inline frankenstein/cmd/frankenstein/main.go:92]: Ah.

**Comment by tomwilkie** [comment]
@juliusv also added some locking to ensure we don't accept anymore samples whilst an ingestor is shutting down.

**Comment by juliusv** [comment]
:+1:

**Pull request merged**